### ISSUE TITLE
off line messages

### ISSUE THREAD

**Issue by zwischen** [irungentoo/toxcore#1187]
I would really appreciate the ability to send an off line message (even if of a limited # of characters) rather than wait for a long-absent contact to return before being able to pass on important information.  I realize I can queue messages for when the contact returns but by then they're often 'stale-dated' before they arrive (and without a date-stamp can also be pointless).

**Comment by aaannndddyyy** [comment]
@dubslow: They must look like and be treated like other dht messages for that. If nodes replicate them and they also have an encrypted TTL which does not differ from the one of other dht messages like the ones that carry the information of id-ip bindings, they can well remain undistinguishable.

**Comment by dfortner** [comment]
Developers said they don't care about offline messaging on IRC and they will probably do something fake like queue messages and wait for both contacts to be online again.
Tox won't be useful without properly implementation of offline messaging that always work. I'm never using it and I hope nobody else does until the developers start seeing what people need.
Where's the security audit, too?
Very weird.

**Comment by fcore117** [comment]
Then dont use, there are Skype for this. I find that if developers even invent true offline messages then it should be switchable to off. Currently faux is best way like Skype 4.x had.

**Comment by ArchangeGabriel** [comment]
@fcore117 I don’t agree. Having true offline messaging would be a very great thing. However, I must agree on the fact that there is a lot of things to think about here, and that this must not be a priority.

While security audit, MITM and things like that should be a top priority. And project organization too.

I also trust Tox more than Skype evidently, because I know for sure that Skype is not preventing M$ and the NSA to read my message. But for Tox, I’m just left to “it might not”, without proof of it. So the only difference currently is that I don’t trust Skype at all and won’t unless it become FOSS and audited, while I might be able to trust Tox when it will have been audited. And also, it currently relies on the fact I exchange ID with my contacts through PGP encrypted mail within my WOT, or IRL. Someone not being that preventive can be MITM’d.

**Comment by aaannndddyyy** [comment]
@dfortner : That's kind of a strong statement, that you hope no one is using it. I am using it, and I agree with you that it is not quite useable for the masses if the masses' needs are not being targeted. But tox is still alpha, and yet it is not all useless even now. IMHO, it has a lot of potential. 
We all wait impatiently for important features like this or for an audit, but otoh things should not be rushed out but rather done well. And an audit at this stage simply doesn't make sense. So no, do not recommend tox for any mission critical use. But even now it has its advantages (and disadvantages) over skype.
So tox even now is not totally useless. And also, I don't know whether devs have said they don't care. If they don't then they, don't. They are not paid developers but do that in their free time. But if they want their project to succeed, I'm sure they won't mind merging a PR implementing that functionality - if done right.
So summing up: I agree that true offline messages are important, more so an independent security audit. I also agree that tox is not (yet) quite useable for the masses. But I don't agree wit the negativeness expressed by hoping that no one uses it. It is still early in development. Only time will tell if tox succeeds or not.

**Comment by bitcard** [comment]
DHT store extension can support the offline message, it support put/get data limit size 1k, please visit http://libtorrent.org/dht_store.html

**Comment by aaannndddyyy** [comment]
@bitcard: 2 hours is a bit low. and tox does use a custom DHT. But thanks for the pointer.

**Comment by autious** [comment]
This is a duplicate of #870

**Comment by bitcard** [comment]
the default is 2 hours, you can set expire time.

**Comment by aaannndddyyy** [comment]
have they solved the spam and dos problem?

**Comment by bitcard** [comment]
spam and dos are general problem in p2p network, should solve it from architecture:first, limit the items size stored on DHT network, second, account limitation, register a account must consume amount of bitcoin or altcoin or other resource. this also can prevent other attack such as sybil attack. so suggest tox import altcoin in tox system.

**Comment by zenware** [comment]
@bitcard requiring some form of registration would mean there is some form of centralization, which is simply not how this system works. Additionally something like a Sybil Attack isn't really possible in this scenario as there is no reputation system, and using some form of altcoin would actually enable a Sybil Attack and add additional complexity.

I do think it would be nice to toss 'offline' messages around the DHT, and I'm not terribly concerned about the content getting there with 100% reliability because I can always send the data again later. But I do think there should be something that would indicate to either the sender or the intended recipient that there was a failure of some sort in the message being delivered. Perhaps some variation of read receipts could be used for that purpose.

I want this project to get into the hands of as many users as possible and I think a lot of people have an expectation for this sort of functionality so for that reason I think it's quite important. Also it's rather technically awesome and I'm interested in researching ways that it could be done.

**Comment by aaannndddyyy** [comment]
> By the way, offline messaging seems to be in the works:
https://github.com/DrakeFish/toxcore/tree/offline_storage

404

**Comment by jf99** [comment]
They stated that DrakeFish vanished at https://blog.tox.chat/2015/08/tox-dev-talks-2/:

"The student who was implementing it disappeared. irungentoo will be taking over if he has time."

**Comment by GrayHatter** [comment]
dup of #186
